hazelcast/src/main/java/com/hazelcast/internal/networking/nio/NioOutboundPipeline.java

```diff
@@ -206,73 +255,118 @@ public void process() throws Exception {
 
         flushToSocket();
 
+        if (migrationRequested()) {
+            startMigration();
+            // we leave this method and the NioOutboundPipeline remains scheduled.
+            // So we don't need to worry about write-through
+            return;
+        }
+
         if (sendBuffer.remaining() > 0) {
             pipelineStatus = DIRTY;
         }
 
         switch (pipelineStatus) {
             case CLEAN:
-                // There is nothing left to be done; so lets unschedule this pipeline
-                unschedule();
+                postProcessClean();
                 break;
             case DIRTY:
-                // pipeline is dirty, so lets register for an OP_WRITE to write
-                // more data.
-                registerOp(OP_WRITE);
+                postProcessDirty();
                 break;
             case BLOCKED:
-                // pipeline is blocked; no point in receiving OP_WRITE events.
-                unregisterOp(OP_WRITE);
+                postProcessBlocked();
                 break;
             default:
                 throw new IllegalStateException();
         }
     }
 
-    /**
-     * Tries to unschedule this pipeline.
-     * <p/>
-     * It will only be unscheduled if:
-     * - there are no pending frames.
-     * <p/>
-     * If the outputBuffer is dirty then it will register itself for an OP_WRITE since we are interested in knowing
-     * if there is more space in the socket output buffer.
-     * If the outputBuffer is not dirty, then it will unregister itself from an OP_WRITE since it isn't interested
-     * in space in the socket outputBuffer.
-     * <p/>
-     * This call is only made by the owning IO thread.
-     */
-    private void unschedule() throws IOException {
-        // since everything is written, we are not interested anymore in write-events, so lets unsubscribe
+    private void postProcessBlocked() throws IOException {
+        // pipeline is blocked; no point in receiving OP_WRITE events.
         unregisterOp(OP_WRITE);
-        // So the outputBuffer is empty, so we are going to unschedule the pipeline.
-        scheduled.set(false);
 
-        if (writeQueue.isEmpty() && priorityWriteQueue.isEmpty()) {
-            // there are no remaining frames, so we are done.
-            return;
+        for (; ; ) {
+            State state = scheduled.get();
+            if (state == State.SCHEDULED || state == State.BLOCKED) {
```

Comment:
1: Yes. At the moment you 'block', it could be that someone tries to wake you up and this needs to lead to a reprocessing of the pipeline. This is needed for TLS when the offloaded hanshake completes when you are trying to block. 

2: both cases need to lead to blocked. It isn't really needed to set blocked if you are already blocked.